### PR TITLE
New docs section to explain how to setup test environment for Custom Applications

### LINKED ISSUE #2569: Document recommended setup for dynamic `entryPointUriPath`
**Which part of the documentation does it affect and which topic does it refer to?**

How to setup the application if the `entryPointUriPath` differs based on environment variables.

This is usually the case if the Custom Application should be configured for "different environments" of the same Custom Application (e.g. dev, staging, prod). 

**Describe what needs to be improved**

This feature is currently not supported (having one app with different deployed versions).

However, there is a workaround to have this kind of setup. It requires to have multiple apps configured that point to different URLs. Therefore, the `entryPointUriPath` can't be "statically" defined  but it must be defined in an environment variable.

We should document how we recommend this setup to be defined in regards to the Custom Application config, to permissions, etc.

### PR DESCRIPTION
**Pull request by CarlosCortizasCT**
#### Summary

The goal of this PR is to add a new section in the documentation to explain to Custom Application developers how they can have a setup which allows to mimic having different application environments.

closes #2569 

#### Description

A new section in the "Help needed?" section is added with instructions on how to apply the required setup.

### CODE REVIEW

File: website/src/content/help-needed.mdx [hunk 1]
```diff
@@ -46,3 +46,67 @@ This implies that changes to the [Custom Application configuration](https://docs
 If you are interested in more advanced functionalities, let us know and open a [support issue](https://github.com/commercetools/merchant-center-application-kit/issues/new/choose).
 
 </Info>
+
+# Using a testing environment
```
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply website/src/content/help-needed.mdx]: Updated in this commit: 8d77495d8009d3b696b225e937ee439a6b55a490

File: website/src/content/help-needed.mdx [hunk 2]
```diff
@@ -46,3 +46,67 @@ This implies that changes to the [Custom Application configuration](https://docs
...
+
+During the development cycle of new features for a Custom Application, it can be helpful to showcase or test those new features before publishing them to the production environment.
+
+Merchant Center platform has only one environment where Custom Applications are installed to, but a similar goal can be achieved by setting up the Custom Application configuration so it uses a different [`entryPointUriPath`](/api-reference/application-config#entrypointuripath) configuration property per its installation deployments.
```
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply website/src/content/help-needed.mdx]: Updated in this commit: 8d77495d8009d3b696b225e937ee439a6b55a490

File: website/src/content/help-needed.mdx [hunk 3]
```diff
@@ -46,3 +46,67 @@ This implies that changes to the [Custom Application configuration](https://docs
...
+
+Let's say there are two deployments of the same application: `A` and `B`. We can assign an environment value for the `entryPointUriPath` property in its configuration.
+
+```json
+{
+  "entryPointUriPath": "${env:ENTRY_POINT_URI_PATH}"
+}
+```
+
+The `ENTRY_POINT_URI_PATH` environment variable can be provided in different ways:
```
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply website/src/content/help-needed.mdx:76]: I tried to rewrite that section a little bit in this commit: 8d77495d8009d3b696b225e937ee439a6b55a490

File: website/src/content/help-needed.mdx [hunk 4]
```diff
@@ -46,3 +46,67 @@ This implies that changes to the [Custom Application configuration](https://docs
...
+// Make sure to import the helper functions from the `ssr` entry point.
+import { entryPointUriPathToPermissionKeys } from '@commercetools-frontend/application-shell/ssr';
+
+export const entryPointUriPath = typeof window === 'undefined' ?
+  process.env.ENTRY_POINT_URI_PATH : window.app.ENTRY_POINT_URI_PATH;
+
+export const PERMISSIONS = entryPointUriPathToPermissionKeys(entryPointUriPath);
+```
+
+The value is read from the environment at build time and from the `window.app` object at runtime (that property is automatically populated by the Custom Application iternals).
```
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply website/src/content/help-needed.mdx]: I rewrote the section in this commit: 8d77495d8009d3b696b225e937ee439a6b55a490

Please let me know what do you think about it.

File: website/src/content/help-needed.mdx [hunk 5]
```diff
@@ -46,3 +46,76 @@ This implies that changes to the [Custom Application configuration](https://docs
...
+  mc-scripts --env .env-A start
+  ```
+
+- by defining the environment variables in your CI service
+
+Using this pattern, different `entryPointUriPath` values can be defined for different environments:
+* `A` environment: `my-application-entry-point`
+* `B` environment: `my-application-entry-point-test`
+
+With this setup, a new preview version of the Custom Application can be deployed to `B` environment and showcasing/testing can be done without affecting the production environment.
```
Comment (emmenko) [inline website/src/content/help-needed.mdx:101]:
```suggestion
With this setup, preview versions of the Custom Application can be deployed to the `B` environment and tested using the application `my-application-entry-point-test`.
```

File: website/src/content/help-needed.mdx [hunk 6]
```diff
@@ -46,3 +46,76 @@ This implies that changes to the [Custom Application configuration](https://docs
...
+// Make sure to import the helper functions from the `ssr` entry point.
+import { entryPointUriPathToPermissionKeys } from '@commercetools-frontend/application-shell/ssr';
+
+export const entryPointUriPath = typeof window === 'undefined' ?
+  process.env.ENTRY_POINT_URI_PATH : window.app.ENTRY_POINT_URI_PATH;
+
+export const PERMISSIONS = entryPointUriPathToPermissionKeys(entryPointUriPath);
+```
+
+The value is read from the environment at build time and from the `window.app` object at runtime (that property is automatically populated by the Custom Application iternals).
```
Comment (emmenko) [inline website/src/content/help-needed.mdx:121]:
````suggestion
However, besides using different `entryPointUriPath` values, the logic of verifying the user permissions in the Custom Application source code must also be adjusted accordingly.

As [explained in this section](/migrating-from-project-level-custom-applications#choose-the-entrypointuripath), the permission names of a Custom Application are derived from the `entryPointUriPath` value. Therefore, using different `entryPointUriPath` values require to change the permission names as well.

We recommend to use the helper function `entryPointUriPathToPermissionKeys` to compute the permission names from the `entryPointUriPath`.

For more information see the [Permissions](/development/permissions) documentation.

<Info>

In this scenario the `entryPointUriPath` value cannot be defined statically but it's injected using an environment variable.

Therefore, reading the value from the source code must take into account both server-side and client-side logic.

```js highlightLines="4-6"
// Make sure to import the helper functions from the `ssr` entry point.
import { entryPointUriPathToPermissionKeys } from '@commercetools-frontend/application-shell/ssr';

export const entryPointUriPath = typeof window === 'undefined'
  ? process.env.ENTRY_POINT_URI_PATH
  : window.app.ENTRY_POINT_URI_PATH;

export const PERMISSIONS = entryPointUriPathToPermissionKeys(entryPointUriPath);
```

</Info>
````

**Review by emmenko — COMMENTED** [review]
Thanks, looks much better now 👌

File: website/src/content/help-needed.mdx (hunk 5 above)
Comment (CarlosCortizasCT) [reply website/src/content/help-needed.mdx]: Updated in this commit: 7bdca6330f1f6e6bc799395e5524ae1f4a2b7f9d

File: website/src/content/help-needed.mdx (hunk 6 above)
Comment (CarlosCortizasCT) [reply website/src/content/help-needed.mdx]: Updated in this commit: 7bdca6330f1f6e6bc799395e5524ae1f4a2b7f9d

Thanks a lot for your help 🙇

File: website/src/content/api-reference/application-config.mdx [hunk 7]
```diff
@@ -165,6 +165,13 @@ Note that the list can be extended at any time. _Last update 2022-01-01_.
 
 </Warning>
 
+<Info>
+
+You can use different `entryPointUriPath` values per deployment to mimic having multiple environments of the same Custom Application.
```
Comment (emmenko) [inline website/src/content/api-reference/application-config.mdx:170]: Nit: I would maybe rephrase this a bit, to avoid wrong expectations.

```suggestion
Even though the `entryPointUriPath` is a unique identifier of a Custom Application, and thus should be defined statically in the source code, it can also be defined "dynamically" via environment variables using the environment placeholders.

This can be the case when having to deploy the Custom Application to different environments.
```

**Review by emmenko — COMMENTED** [review]
Last improvement suggestion, otherwise looks good.

File: website/src/content/api-reference/application-config.mdx (hunk 7 above)
Comment (CarlosCortizasCT) [reply website/src/content/api-reference/application-config.mdx:170]: Perhaps we want to link the "using environment placeholders" to [the section we explain this part](http://localhost:8000/api-reference/application-config#using-variable-placeholders)?

Comment (emmenko) [reply website/src/content/api-reference/application-config.mdx:170]: Sure, good idea 👌

Comment (CarlosCortizasCT) [reply website/src/content/api-reference/application-config.mdx]: Updated in this commit: f3ad4c9d5cf870b5787c36dcf77634b9f88b7051

**Review by emmenko — APPROVED** [review]
💯

File: website/src/content/help-needed.mdx [hunk 8]
```diff
@@ -46,3 +46,82 @@ This implies that changes to the [Custom Application configuration](https://docs
...
+
+# I want to use a test or staging environment for my Custom Application
+
+During the development cycle of new features for a Custom Application, it can be helpful to showcase or test those new features before publishing them to the production environment.
+
+This generally can be achieved using preview deployments, where you can inspect different versions of the Custom Application. As of now, the Merchant Center does not support this feature yet.
+
+The only option to emulate this setup is to create multiple Custom Applications in the Merchant Center, each one with a different [`entryPointUriPath`](/api-reference/application-config#entrypointuripath).
+
+However, the Custom Application must be configured to support "dynamic" values for the `entryPointUriPath` and build different versions of it.
```
Comment (adinakleine) [inline website/src/content/help-needed.mdx:58]:
```suggestion
However, the Custom Application must be configured to support dynamic values for the `entryPointUriPath` and build different versions of it.
```

**Review by adinakleine — APPROVED** [review]
Thanks @CarlosCortizasCT 🙌 . Find some suggestions... mainly on writing style again. ✍️

File: website/src/content/help-needed.mdx [hunk 9]
```diff
@@ -46,3 +46,82 @@ This implies that changes to the [Custom Application configuration](https://docs
...
+
+<Info>
+
+Custom Applications "dynamic" configurations (per environment) setup are further explained in [this section](/api-reference/application-config#environment-variable-references).
```
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply website/src/content/help-needed.mdx:62]: I was wondering if something like this could also work:
```
Setting up dynamic configurations of Custom Applications (per environment) is further...
```

File: website/src/content/api-reference/application-config.mdx [hunk 10]
```diff
@@ -165,6 +165,16 @@ Note that the list can be extended at any time. _Last update 2022-01-01_.
 
 </Warning>
 
+<Info>
+
+Even though the `entryPointUriPath` is a unique identifier of a Custom Application, and thus should be defined statically in the source code, it can also be defined "dynamically" via environment variables using the [environment placeholders](#using-variable-placeholders).
```
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply website/src/content/api-reference/application-config.mdx]: Thanks for your help @adinakleine 👍 

All suggestions have been applied in this commit: b0a9d85dc886a0ec6c21ae1e209bf28407ba57c7

(I opted for skipping the quotation as suggested)

**Comment by emmenko** [comment]
Thanks! 🥳
